server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -1480,16 +1478,19 @@ private void applyGlobalStateRestore(ClusterState currentState, Metadata.Builder
                     mdBuilder.put(cursor);
                 }
             }
+
+            // override existing restorable customs (as there might be noting in snapshot to override them)
+            mdBuilder.removeCustomIf((key, value) -> value.isRestorable());
+
+            // restore customs from the snapshot
             if (metadata.customs() != null) {
-                for (Map.Entry<String, Metadata.Custom> cursor : metadata.customs().entrySet()) {
-                    if (RepositoriesMetadata.TYPE.equals(cursor.getKey()) == false
-                        && DataStreamMetadata.TYPE.equals(cursor.getKey()) == false
-                        && cursor.getValue() instanceof Metadata.NonRestorableCustom == false) {
+                for (var entry : metadata.customs().entrySet()) {
+                    if (entry.getValue().isRestorable()) {
```

Comment:
This part is tested in SystemDataStreamSnapshotIT and AutoscalingSnapshotsIT. I am working on creating a IT for the scenario mentioned in the pr description.

Reply:
This behavior is also covered by CustomMetadataSnapshotIT